cpp/src/io/json/read_json.cu

```diff
@@ -89,36 +90,35 @@ device_span<char> ingest_raw_input(device_span<char> buffer,
       std::upper_bound(prefsum_source_sizes.begin(), prefsum_source_sizes.end(), range_offset);
     size_t start_source = std::distance(prefsum_source_sizes.begin(), upper);
 
-    auto remaining_bytes_to_read = std::min(range_size, prefsum_source_sizes.back() - range_offset);
+    auto total_bytes_to_read = std::min(range_size, prefsum_source_sizes.back() - range_offset);
     range_offset -= start_source ? prefsum_source_sizes[start_source - 1] : 0;
-    for (size_t i = start_source; i < sources.size() && remaining_bytes_to_read; i++) {
+    for (size_t i = start_source; i < sources.size() && bytes_read < total_bytes_to_read; i++) {
       if (sources[i]->is_empty()) continue;
-      auto data_size   = std::min(sources[i]->size() - range_offset, remaining_bytes_to_read);
-      auto destination = reinterpret_cast<uint8_t*>(buffer.data()) + bytes_read;
+      auto data_size =
+        std::min(sources[i]->size() - range_offset, total_bytes_to_read - bytes_read);
+      auto destination = reinterpret_cast<uint8_t*>(buffer.data()) + bytes_read +
+                         (num_delimiter_chars * delimiter_map.size());
       if (sources[i]->is_device_read_preferred(data_size)) {
         bytes_read += sources[i]->device_read(range_offset, data_size, destination, stream);
       } else {
-        h_buffers.emplace_back(sources[i]->host_read(range_offset, data_size));
-        auto const& h_buffer = h_buffers.back();
+        auto h_buffer = sources[i]->host_read(range_offset, data_size);
         CUDF_CUDA_TRY(cudaMemcpyAsync(
-          destination, h_buffer->data(), h_buffer->size(), cudaMemcpyDefault, stream.value()));
+          destination, h_buffer->data(), h_buffer->size(), cudaMemcpyHostToDevice, stream.value()));
```

Comment:
`h_buffer` is destroyed right away but `stream` is not yet sync. To avoid sync every iteration, we need to store `h_buffer` into some temp array so they can outlive until the end of this function.

Reply:
See example in https://github.com/rapidsai/cudf/blob/b810113d6255dbe123aafbc80018bf6165a0842f/cpp/src/io/orc/reader_impl_chunking.cu#L515-L522.

Reply:
Oh I just realized that such temp array is removed in this code (`h_buffers`). Why?

Reply:
We don't need to store the host buffers into an array since `host_read` stores the buffer in pageable memory, making `cudaMemcpyAsync` a blocking operation.

Reply:
```
  /**
   * @brief Returns a buffer with a subset of data from the source.
   *
   * @param[in] offset Bytes from the start
   * @param[in] size Bytes to read
   *
   * @return The data buffer (can be smaller than size)
   */
  virtual std::unique_ptr<datasource::buffer> host_read(size_t offset, size_t size) = 0;
```
I don't see any requirement for the `host_read` function to be pageable. So I assume that such situation is outlier and cannot be guaranteed.

Reply:
I missed this. We're usually explicit about synchronization and don't rely on pageable copies being sync. I would suggest a `stream.synchronize()` after the copy.

Reply:
I would object to that since we will sync at every iteration, and lean toward storing all the buffers then sync only once.

Reply:
hmm, that is only relevant when reading from a pinned host buffer... which is exactly how Spark uses the libcudf readers (afaik). Yeah, it's better to keep all temp buffers :+1:

Reply:
Thank you, that's a good point. The reasoning here was that using pinned memory for `host_read` is unlikely since (i) the pool size will be smaller than the total data size, especially with multi-source reads, and (ii) new pinned memory allocations is going to be expensive in the case we have to read multiple whole files.